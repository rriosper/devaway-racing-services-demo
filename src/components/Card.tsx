import styled from '@xstyled/styled-components';
import {
  alignItems,

  backgroundColor, gridGap,
  gridTemplateColumns,
  gridTemplateRows,
  justifyContent,
  margin,
  padding,
} from '@xstyled/system';
import React, { useMemo } from 'react';

import { ThemeSchema } from '../theme';

// Card
type CardProps = {
  padding?: ThemeSchema['space'];
  margin?: ThemeSchema['space'];
  gridTemplateColumns?: string;
  gridTemplateRows?: string;
  gridGap?: string;
  justifyContent?: string;
  alignItems?: string;
  backgroundColor?: ThemeSchema['colors'];
};

export const Card = styled.div<CardProps>`
  ${padding}
  ${margin}
  ${gridTemplateColumns}
  ${gridTemplateRows}
  ${gridGap}
  ${justifyContent}
  ${alignItems}
  ${backgroundColor}
  background-color: card;
  box-shadow: card;
  display: grid;
  border-radius: card;
`;

Card.defaultProps = {
  padding: 'md',
};

// CardField
type CardFieldsProps = {
  fields: Array<{
    label: string;
    value: Nullable<string | number>;
  }>;
};

const SCardFieldsWrapper = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-row-gap: md;
  grid-column-gap: sm;
`;

const SCardField = styled.div`
  display: flex;
  flex-direction: column;

  > div {
    font-weight: bold;
    color: foreground;
    text-trasnform: lowercase;
    margin-bottom: xs;

    :first-letter {
      text-transform: uppercase;
    }
  }
`;

export const CardFields: React.FC<CardFieldsProps> = ({ fields }) => {
  const fieldsComponent = useMemo(
    () => fields.map(({ label, value }) => (
      <SCardField key={label}>
        <div>{label}</div>
        <span>{value}</span>
      </SCardField>
    )),
    [fields],
  );
  return <SCardFieldsWrapper>{fieldsComponent}</SCardFieldsWrapper>;
};
