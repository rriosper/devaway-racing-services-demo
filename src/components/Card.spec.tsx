import React from 'react';
import { render } from '@testing-library/react';

import { Card, CardFields } from './Card';

test('Card', () => {
  const { asFragment, getByText } = render(<Card>CHILDREN</Card>);
  expect(getByText('CHILDREN')).toBeInTheDocument();
  expect(asFragment).toMatchSnapshot();
});

test('CardFields', () => {
  const { asFragment, getByText } = render(
    <CardFields
      fields={[
        { label: 'Label 1', value: 'Value 1' },
        { label: 'Label 2', value: 'Value 2' },
      ]}
    />,
  );

  expect(getByText('Label 1')).toBeInTheDocument();
  expect(getByText('Value 1')).toBeInTheDocument();
  expect(getByText('Label 2')).toBeInTheDocument();
  expect(getByText('Value 2')).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});
