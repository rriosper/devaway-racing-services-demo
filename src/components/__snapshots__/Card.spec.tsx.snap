// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 1`] = `[Function]`;

exports[`CardFields 1`] = `
<DocumentFragment>
  .c0 {
  display: grid;
  grid-template-columns: repeat(2,1fr);
  grid-row-gap: md;
  grid-column-gap: sm;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 > div {
  font-weight: bold;
  color: foreground;
  text-trasnform: lowercase;
  margin-bottom: xs;
}

.c1 > div:first-letter {
  text-transform: uppercase;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        Label 1
      </div>
      <span>
        Value 1
      </span>
    </div>
    <div
      class="c1"
    >
      <div>
        Label 2
      </div>
      <span>
        Value 2
      </span>
    </div>
  </div>
</DocumentFragment>
`;
